### ISSUE TITLE
celery_executor becomes stuck if child process receives signal before reset_signals is called

### ISSUE THREAD

**Issue by yuqian90** [apache/airflow#15938]
**Apache Airflow version**: 1.10.13 onwards (Any version that picked up #11278, including Airflow 2.0.* and 2.1.*)

**Environment**:
- **Cloud provider or hardware configuration**: Any
- **OS** (e.g. from /etc/os-release): Only tested on Debian Linux, but others may be affected too
- **Kernel** (e.g. `uname -a`): Any
- **Install tools**: Any
- **Others**: Only celery_executor is affected

**What happened**:
This was first reported [here](https://github.com/apache/airflow/issues/7935#issuecomment-839656436).
airflow-scheduler sometimes stops heartbeating and stops scheduling any tasks with this last line in the log. This happen at random times, a few times a week. Happens more often if the scheduler machine is slow. 

```
{scheduler_job.py:746} INFO - Exiting gracefully upon receiving signal 15
```

The problem is that sometimes the machine is slow, `reset_signals()` of one or more slow child processes is not yet called before other child processes send `SIGTERM` when they exit. As a result, the slow child processes respond to the `SIGTERM` as if they are the main scheduler process. Thus we see the `Exiting gracefully upon receiving signal 15` in the scheduler log. Since the probability of this happening is very low, this issue is really difficult to reproduce reliably in production.

Related to #7935
Most likely caused by #11278

**What you expected to happen**: 
Scheduler should not become stuck

**How to reproduce it**:

Here's a small reproducing example of the problem. There's roughly 1/25 chance it will be stuck. Run it many times to see it happen.

```python
#!/usr/bin/env python3.8
import os
import random
import signal
import time
from multiprocessing import Pool


def send_task_to_executor(arg):
    pass


def _exit_gracefully(signum, frame):
    print(f"{os.getpid()} Exiting gracefully upon receiving signal {signum}")


def register_signals():
    print(f"{os.getpid()} register_signals()")
    signal.signal(signal.SIGINT, _exit_gracefully)
    signal.signal(signal.SIGTERM, _exit_gracefully)
    signal.signal(signal.SIGUSR2, _exit_gracefully)


def reset_signals():
    if random.randint(0, 500) == 0:
        # This sleep statement here simulates the machine being busy
        print(f"{os.getpid()} is slow")
        time.sleep(0.1)
    signal.signal(signal.SIGINT, signal.SIG_DFL)
    signal.signal(signal.SIGTERM, signal.SIG_DFL)
    signal.signal(signal.SIGUSR2, signal.SIG_DFL)


if __name__ == "__main__":
    register_signals()

    task_tuples_to_send = list(range(20))
    sync_parallelism = 15
    chunksize = 5

    with Pool(processes=sync_parallelism, initializer=reset_signals) as pool:
        pool.map(
            send_task_to_executor,
            task_tuples_to_send,
            chunksize=chunksize,
        )


```

The reproducing example above can become stuck with a `py-spy dump` that looks exactly like what airflow scheduler does:

`py-spy dump` for the parent `airflow scheduler` process
```
Python v3.8.7

Thread 0x7FB54794E740 (active): "MainThread"
    poll (multiprocessing/popen_fork.py:27)
    wait (multiprocessing/popen_fork.py:47)
    join (multiprocessing/process.py:149)
    _terminate_pool (multiprocessing/pool.py:729)
    __call__ (multiprocessing/util.py:224)
    terminate (multiprocessing/pool.py:654)
    __exit__ (multiprocessing/pool.py:736)
    _send_tasks_to_celery (airflow/executors/celery_executor.py:331)
    _process_tasks (airflow/executors/celery_executor.py:272)
    trigger_tasks (airflow/executors/celery_executor.py:263)
    heartbeat (airflow/executors/base_executor.py:158)
    _run_scheduler_loop (airflow/jobs/scheduler_job.py:1388)
    _execute (airflow/jobs/scheduler_job.py:1284)
    run (airflow/jobs/base_job.py:237)
    scheduler (airflow/cli/commands/scheduler_command.py:63)
    wrapper (airflow/utils/cli.py:89)
    command (airflow/cli/cli_parser.py:48)
    main (airflow/__main__.py:40)
    <module> (airflow:8)
```

`py-spy dump` for the child `airflow scheduler` process

```
Python v3.8.7

Thread 16232 (idle): "MainThread"
    __enter__ (multiprocessing/synchronize.py:95)
    get (multiprocessing/queues.py:355)
    worker (multiprocessing/pool.py:114)
    run (multiprocessing/process.py:108)
    _bootstrap (multiprocessing/process.py:315)
    _launch (multiprocessing/popen_fork.py:75)
    __init__ (multiprocessing/popen_fork.py:19)
    _Popen (multiprocessing/context.py:277)
    start (multiprocessing/process.py:121)
    _repopulate_pool_static (multiprocessing/pool.py:326)
    _repopulate_pool (multiprocessing/pool.py:303)
    __init__ (multiprocessing/pool.py:212)
    Pool (multiprocessing/context.py:119)
    _send_tasks_to_celery (airflow/executors/celery_executor.py:330)
    _process_tasks (airflow/executors/celery_executor.py:272)
    trigger_tasks (airflow/executors/celery_executor.py:263)
    heartbeat (airflow/executors/base_executor.py:158)
    _run_scheduler_loop (airflow/jobs/scheduler_job.py:1388)
    _execute (airflow/jobs/scheduler_job.py:1284)
    run (airflow/jobs/base_job.py:237)
    scheduler (airflow/cli/commands/scheduler_command.py:63)
    wrapper (airflow/utils/cli.py:89)
    command (airflow/cli/cli_parser.py:48)
    main (airflow/__main__.py:40)
    <module> (airflow:8)
```

**Comment by potiuk** [comment]
Whoa!

**Comment by ashb** [comment]
Just _how_ slow does it have to be to happen?

Nice digging.

**Comment by ashb** [comment]
We can probably guard this by closing of the current pid when we register them, and checking that the signal is received by the same pid

**Comment by gdevanla** [comment]
I just upgraded to 2.0.2 and this happens almost within 1 minute of scheduler restart.  We are on Python 3.8 and using RabbitMq/Celery.

```
[2021-05-21 05:40:15,996] {channel.py:446} DEBUG - Channel open
[2021-05-21 05:40:16,000] {channel.py:105} DEBUG - using channel_id: 1
[2021-05-21 05:40:16,005] {channel.py:446} DEBUG - Channel open
[[2021-05-21 05:40:16,053] {2021-05-21 05:40:16,053scheduler_job.py:] {746} INFOscheduler_job.py: - Exiting gracefully upon receiving signal 15746} INFO
 - Exiting gracefully upon receiving signal 15
[2021-05-21 05:40:16,060] {scheduler_job.py:746} INFO - Exiting gracefully upon receiving signal 15
[2021-05-21 05:40:16,067] {scheduler_job.py:746} INFO - Exiting gracefully upon receiving signal 15
[2021-05-21 05:40:16,076] {scheduler_job.py:746} INFO - Exiting gracefully upon receiving signal 15
[2021-05-21 05:43:34,715] {serialized_dag.py:197} DEBUG - Deleting Serialized DAGs (for which DAG files are deleted) from serialized_dag table 
[2021-05-21 05:43:34,716] {dag.py:2213} DEBUG - Deactivating DAGs (for which DAG files are deleted) from dag table 
[2021-05-21 05:43:34,723] {dagcode.py:135} DEBUG - Deleting code from dag_code table 
[2021-05-21 05:43:48,746] {settings.py:292} DEBUG - Disposing DB connection pool (PID 6758)
[2021-05-21 05:43:48,760] {scheduler_job.py:310} DEBUG - Waiting for <ForkProcess name='DagFileProcessor1-Process' pid=6758 parent=1204 stopped exitcode=0>
```

And here is the corresponding logs in the `error` file: (I am not sure if this is related, but just thought I would add this if it could be useful)

```
Process ForkPoolWorker-622:
Traceback (most recent call last):
  File "/home/imcron/py_src/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/home/imcron/py_src/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/imcron/py_src/lib/python3.8/multiprocessing/pool.py", line 114, in worker
    task = get()
  File "/home/imcron/py_src/lib/python3.8/multiprocessing/queues.py", line 355, in get
    with self._rlock:
  File "/home/imcron/py_src/lib/python3.8/multiprocessing/synchronize.py", line 95, in __enter__
    return self._semlock.__enter__()
  File "/home/imcron/rapc/.env38/lib/python3.8/site-packages/airflow/jobs/scheduler_job.py", line 748, in _exit_gracefully
    self.processor_agent.end()
  File "/home/imcron/rapc/.env38/lib/python3.8/site-packages/airflow/utils/dag_processing.py", line 458, in end
    self._process.join(timeout=1.0)
  File "/home/imcron/py_src/lib/python3.8/multiprocessing/process.py", line 147, in join
    assert self._parent_pid == os.getpid(), 'can only join a child process'
AssertionError: can only join a child process
```

**Comment by gdevanla** [comment]
On a related note, we used to have `-r` option to recycle the scheduler.  Looks like that option is not available anymore. That option would have been a nice workaround for this kind of issue.

**Comment by yuqian90** [comment]
> Just how slow does it have to be to happen?
> We can probably guard this by closing of the current pid when we register them, and checking that the signal is received by the same pid

Hi, @ashb it's not clear to me how slow it must be exactly for this to happen. It looks like as long as some child processes are a fraction of a second slower than the others, they easily get into a deadlock when a SIGTERM is received. So even a transient slowness of a beefy machine can cause this to happen. 

Here's what I tried so far. Only the last method seems to fix the issue completely (i.e. we have to stop using `multiprocessing.Pool`):
- Tried to reset the signal handler to `signal.SIG_DFL` in `register_signals` if the current process is a child process. This doesn't help because the child process inherits the parent's signal handler when it's forked. Still hangs occasionally.
- Tried to make `_exit_gracefully` a no-op if the current process is a child process. This isn't sufficient. Still hangs occasionally.
- Tried to change multiprocessing to use "spawn" instead of "fork" like some people suggested [on the internet](https://pythonspeed.com/articles/python-multiprocessing/), it greatly reduced the chance of this issue happening. However, after running the reproducing example about 8000 times, it still happened. So it doesn't fix the issue completely.
- **Replace `multiprocessing.Pool` with `concurrent.futures.process.ProcessPoolExecutor`. Once this is done, the reproducing example no longer hangs even after running it tens of thousands times.**. So I put up PR #15938 which fixes the issue using this method. 

From experience, `multiprocessing.Pool` is notorious for causing mysterious hangs like these. Using `ProcessPoolExecutor` does not cause the same problems. It has similar interface and uses similar underlying libraries. I don't understand exactly why it fixes the issue, but in practice it always seems to help.

**Comment by ashb** [comment]
I have _no_ problem with us not using multiprocessing - this isn't or first such problem like this

**Comment by BwL1289** [comment]
commenting to keep UTD on this issue

**Comment by thesuperzapper** [comment]
I just wanted to share that the [User-Community Airflow Helm Chart](https://github.com/airflow-helm/charts/tree/main/charts/airflow) now has a mitigation for this issue that will automatically restart the scheduler if no tasks are created within some threshold time.

It's called the scheduler ["Task Creation Check"](https://github.com/airflow-helm/charts/blob/main/charts/airflow/docs/faq/monitoring/scheduler-liveness-probe.md), but it's not enabled by default as, the "threshold" must be longer than your shorted DAG `schedule_interval`, which we don't know unless the user tells us.

**Comment by chuxiangfeng** [comment]
> I just wanted to share that the [User-Community Airflow Helm Chart](https://github.com/airflow-helm/charts/tree/main/charts/airflow) now has a mitigation for this issue that will automatically restart the scheduler if no tasks are created within some threshold time.
> 
> It's called the scheduler ["Task Creation Check"](https://github.com/airflow-helm/charts/blob/main/charts/airflow/docs/faq/monitoring/scheduler-liveness-probe.md), but it's not enabled by default as, the "threshold" must be longer than your shorted DAG `schedule_interval`, which we don't know unless the user tells us.

hello，if i don`t want scheduler restart，how can i do？

**Comment by chuxiangfeng** [comment]
> I just wanted to share that the [User-Community Airflow Helm Chart](https://github.com/airflow-helm/charts/tree/main/charts/airflow) now has a mitigation for this issue that will automatically restart the scheduler if no tasks are created within some threshold time.
> 
> It's called the scheduler ["Task Creation Check"](https://github.com/airflow-helm/charts/blob/main/charts/airflow/docs/faq/monitoring/scheduler-liveness-probe.md), but it's not enabled by default as, the "threshold" must be longer than your shorted DAG `schedule_interval`, which we don't know unless the user tells us.

i hava install a new airflow by helm chart， now no jobs and works，scheduler restart always restart， if i don`t want scheduler restart，how can i do？

**Comment by zsdyx** [comment]
I have the same problem. In version 2.4.1,  schedule pod will restart once in five minutes, I don't know if this behavior is normal. Is it because of the protection mechanism?

**Comment by potiuk** [comment]
> I have the same problem. In version 2.4.1, schedule pod will restart once in five minutes, I don't know if this behavior is normal. Is it because of the protection mechanism?

You need to check why it is restarting. This is not normal. Check your Airflow and K8s logs why it happens.

**Comment by zsdyx** [comment]
> > I have the same problem. In version 2.4.1, schedule pod will restart once in five minutes, I don't know if this behavior is normal. Is it because of the protection mechanism?
> 
> You need to check why it is restarting. This is not normal. Check your Airflow and K8s logs why it happens.

Nodes don't have any error logs.What happens to this problem

https://files.slack.com/files-pri/T02G2NC2F-F04E3UP83GA/screen_shot_2022-12-07_at_17.24.11.png
https://files.slack.com/files-pri/T02G2NC2F-F04E3RQSC6P/screen_shot_2022-12-07_at_17.25.34.png

**Comment by potiuk** [comment]
Those links are behind a login. so :shrug: 

And  think it makes rather little sense to to try to "debug" your issue in a closed issue that has similar message. If you have a problem with restarting nodes, this is likely problem with your deployment, not Airflow - so I strongly suggest you try to analyse your deployment.

BTW. If your logs are not showing anything suspicious and description of the problem is "my nodes are restarting" it makes no sense for anyone else to look at this. I guess you understtand that you won't get any help unless there is something that can be diagnosed? People here try to help in their free time with your issues, so it's on your shoulders to get as much useful  information if you need help with it, I am afraid.e

**Comment by zsdyx** [comment]
Provide some important logs first. After the service is start, Output after each interval.
```
Exiting gracefully upon receiving signal 15
 Sending Signals.SIGTERM to group 20. PIDs of all processes in the group: [20]

```

ENV
GCP: k8s
python: 3.10
airflow: 2.4.1
excutor:CeleryExecutor
```
INFO,"	<TaskInstance: vi_base_schema_scan.vi_base_schema_scan scheduled__2022-12-08T15:45:00+00:00 [running]> in state STARTED",2022-12-09T01:54:53.346995938Z
INFO,"	<TaskInstance: aa_table_lineage.aa_table_lineage scheduled__2022-12-08T21:00:00+00:00 [running]> in state STARTED",2022-12-09T01:54:53.346992107Z
INFO,"	<TaskInstance: vi_base_table_lineage.vi_base_table_lineage scheduled__2022-12-08T19:50:00+00:00 [running]> in state STARTED",2022-12-09T01:54:53.346986570Z
INFO,[2022-12-09T01:54:53.346+0000] {celery_executor.py:534} INFO - Adopted the following 128 tasks from a dead executor,2022-12-09T01:54:53.346933893Z
INFO,[2022-12-09T01:54:52.470+0000] {settings.py:543} INFO - Loaded airflow_local_settings from /opt/airflow/config/airflow_local_settings.py .,2022-12-09T01:54:52.470955248Z
INFO,[2022-12-09T01:54:52.468+0000] {settings.py:58} INFO - Configured default timezone Timezone('UTC'),2022-12-09T01:54:52.469795791Z
INFO,[2022-12-09T01:54:52.465+0000] {scheduler_job.py:1374} INFO - Resetting orphaned tasks for active dag runs,2022-12-09T01:54:52.466143332Z
INFO,[2022-12-09T01:54:52.462+0000] {manager.py:163} INFO - Launched DagFileProcessorManager with pid: 20,2022-12-09T01:54:52.465243117Z
INFO,[2022-12-09T01:54:52.450+0000] {executor_loader.py:107} INFO - Loaded executor: CeleryExecutor,2022-12-09T01:54:52.451308381Z
INFO,[2022-12-09T01:54:52.258+0000] {scheduler_job.py:706} INFO - Processing each file at most -1 times,2022-12-09T01:54:52.259176063Z
INFO,[2022-12-09T01:54:52.258+0000] {scheduler_job.py:701} INFO - Starting the scheduler,2022-12-09T01:54:52.258964320Z
INFO, _/_/  |_/_/  /_/    /_/    /_/  \____/____/|__/,2022-12-09T01:54:52.258959593Z
INFO,___  ___ |  / _  /   _  __/ _  / / /_/ /_ |/ |/ /,2022-12-09T01:54:52.258955844Z
INFO,____  /| |_  /__  ___/_  /_ __  /_  __ \_ | /| / /,2022-12-09T01:54:52.258952230Z
INFO, ____    |__( )_________  __/__  /________      __,2022-12-09T01:54:52.258946962Z
INFO,  ____________       _____________,2022-12-09T01:54:52.258915119Z
INFO,,2022-12-09T01:54:48.293564892Z
INFO,[2022-12-09T01:54:43.752+0000] {scheduler_job.py:775} INFO - Exited execute loop,2022-12-09T01:54:43.752660415Z
INFO,[2022-12-09T01:54:43.752+0000] {process_utils.py:98} INFO - Sending the signal Signals.SIGTERM to process 20 as process group is missing.,2022-12-09T01:54:43.752538440Z
INFO,[2022-12-09T01:54:43.752+0000] {process_utils.py:84} INFO - Sending the signal Signals.SIGTERM to group 20,2022-12-09T01:54:43.752411066Z
INFO,[2022-12-09T01:54:43.752+0000] {process_utils.py:129} INFO - Sending Signals.SIGTERM to group 20. PIDs of all processes in the group: [],2022-12-09T01:54:43.752361863Z
INFO,"[2022-12-09T01:54:43.701+0000] {process_utils.py:79} INFO - Process psutil.Process(pid=20, status='terminated', exitcode=0, started='01:49:51') (20) terminated with exit code 0",2022-12-09T01:54:43.702040509Z
INFO,[2022-12-09T01:54:43.529+0000] {process_utils.py:84} INFO - Sending the signal Signals.SIGTERM to group 20,2022-12-09T01:54:43.529318886Z
INFO,[2022-12-09T01:54:43.528+0000] {process_utils.py:129} INFO - Sending Signals.SIGTERM to group 20. PIDs of all processes in the group: [20],2022-12-09T01:54:43.529225207Z
INFO,[2022-12-09T01:54:42.526+0000] {scheduler_job.py:173} INFO - Exiting gracefully upon receiving signal 15,2022-12-09T01:54:42.527221527Z
```

**Comment by potiuk** [comment]
@zsdyx Can you please open a new issue for it with all the details in it - exactly what happens for you , what is the sequence of events which version, what deployment and all the logs ? 

It's extremely hard to understand if the problem you have is related or not to this closed issue. There is no easy reproduction steps because we have no idea what you have done to get to the state you got there,  and how close it matches that other, already fixed issue. Just spend a bit more time in explaining your situation in detail - help us to help to solve the problem you have please.
